bigframes/core/node_align.py

```diff
@@ -0,0 +1,167 @@
+            )
+        elif isinstance(
+            curr_node,
+            (
+                bigframes.core.nodes.WindowOpNode,
+                bigframes.core.nodes.PromoteOffsetsNode,
+            ),
+        ):
+            pass
+        else:
+            return ExpressionSpec(expression, curr_node)
+        curr_node = curr_node.child
+
+
+def _linearize_trees(
+    base_tree: bigframes.core.nodes.BigFrameNode,
+    append_tree: bigframes.core.nodes.BigFrameNode,
+) -> bigframes.core.nodes.BigFrameNode:
+    """Linearize two divergent tree who only diverge through different additive nodes."""
+    assert append_tree.projection_base == base_tree.projection_base
+    if append_tree == append_tree.projection_base:
+        return base_tree
+    else:
+        assert isinstance(append_tree, ADDITIVE_NODES)
+        return append_tree.replace_child(_linearize_trees(base_tree, append_tree.child))
+
+
+def combine_nodes(
+    l_node: bigframes.core.nodes.BigFrameNode,
+    r_node: bigframes.core.nodes.BigFrameNode,
+) -> bigframes.core.nodes.BigFrameNode:
+    assert l_node.projection_base == r_node.projection_base
+    l_node, l_selection = pull_up_selection(l_node)
+    r_node, r_selection = pull_up_selection(
+        r_node, rename_vars=True
+    )  # Rename only right vars to avoid collisions with left vars
+    combined_selection = (*l_selection, *r_selection)
+    merged_node = _linearize_trees(l_node, r_node)
+    return bigframes.core.nodes.SelectionNode(merged_node, combined_selection)
+
+
+def join_as_projection(
+    l_node: bigframes.core.nodes.BigFrameNode,
+    r_node: bigframes.core.nodes.BigFrameNode,
+    join_keys: Tuple[Tuple[str, str], ...],
+) -> Optional[bigframes.core.nodes.BigFrameNode]:
+    """Joins the two nodes"""
+    if l_node.projection_base != r_node.projection_base:
+        return None
+    # check join key
+    for l_key, r_key in join_keys:
+        # Caller is block, so they still work with raw strings rather than ids
+        left_id = bigframes.core.identifiers.ColumnId(l_key)
+        right_id = bigframes.core.identifiers.ColumnId(r_key)
+        if get_expression_spec(l_node, left_id) != get_expression_spec(
+            r_node, right_id
+        ):
+            return None
+    return combine_nodes(l_node, r_node)
+
+
+def pull_up_selection(
+    node: bigframes.core.nodes.BigFrameNode, rename_vars: bool = False
+) -> Tuple[
+    bigframes.core.nodes.BigFrameNode,
+    Tuple[
+        Tuple[bigframes.core.expression.DerefOp, bigframes.core.identifiers.ColumnId],
+        ...,
+    ],
+]:
+    """Remove all selection nodes above the base node. Returns stripped tree."""
+    if node == node.projection_base:  # base case
+        return node, tuple(
+            (bigframes.core.expression.DerefOp(field.id), field.id)
+            for field in node.fields
+        )
+    assert isinstance(node, (bigframes.core.nodes.SelectionNode, *ADDITIVE_NODES))
```

Comment:
Am I remembering correctly that `SelectionNode` represents a WHERE clause, based on [Selection from relational algebra terminology](https://en.wikipedia.org/wiki/Selection_(relational_algebra))?

Given the high-level that implicit joiner no longer handles these, I'm a little confused why we need to rewrite these? Is it so that they are all consolidated so that the nodes we can combine appear together in the tree? I would appreciate a bit more info.

Reply:
Its actually poorly named. SelectionNode just renames/reorders existing columns, without any scalar transforms